[refactor] simplify expressions

asolntsev opened the pull request:
this commit mostly simplifies "equals" and "hashCode" implementations by using Java built-in methods `Objects.equals` and `Objects.hash`.

Motivation:

Improve Netty codebase.

Modification:

Small refactorings that simplify code and don't affect functionality.

Result:

Slightly improved code: simplified IFs, shortened methods.

asolntsev committed:
[refactor] simplify expressions

this commit mostly simplifies "equals" and "hashCode" implementations by using Java built-in methods `Objects.equals` and `Objects.hash`.

doom369 commented:
This will impact performance because `Objects.hash()` uses varargs == allocation + conversion of all primitives into wrappers

asolntsev commented:
> This will impact performance because `Objects.hash()` uses varargs == allocation + conversion of all primitives into wrappers

@doom369 Can you share some proof for that? Is there really an measurable impact on performance?

(If yes, then we can easily overcome it by creating few overloaded alternatives for `Objects.hash` with 2, 3, 4 and 5 parameters)

chrisvest commented:
Just revert the hashcode changes. Is highly plausible that the array allocation and autoboxing adds overhead. We're not gonna check every case that it doesn't.

Netty is very widely used and many projects and systems rely on its performance, so details like this matter.

asolntsev committed:
avoid using method `java.lang.Objects#hash`

... because it creates a new array which might affect performance.

Now we use overloaded methods `io.netty.util.internal.ObjectUtil.hash*` that don't use varargs, thus not creating arrays.

asolntsev commented:
@doom369 @chrisvest I've removed all usages of `java.lang.Objects.hash*`. 
Now this code is still simple, but doesn't cause array allocation and autoboxing. No overhead.

doom369 commented on codec-base/src/main/java/io/netty/handler/codec/DefaultHeaders.java:
```diff
@@ -988,8 +988,8 @@ public final int hashCode(HashingStrategy<V> valueHashingStrategy) {
             List<V> values = getAll(name);
-            for (int i = 0; i < values.size(); ++i) {
-                result = 31 * result + valueHashingStrategy.hashCode(values.get(i));
+            for (V value : values) {
```
iterator allocation introduced. you need to keep get(i), please revert this and similar changes. Also this PR is very big with many different refactorings. I recommend you to split it into multiple smaller one, 1 pr - 1 kind of refactorings so it's easier to review.

asolntsev replied:
@doom369 I've reverted all changes with for loops. No iterator allocations anymore.

P.S. Generally I agree that PR could be split to 2-3 smaller commits. 

BUT
The PR may seem big in terms of LOC. But it's not big in terms of "different refactorings". These all are very simple, automated refactorings. Mostly simplification of "equals" and "hashCode" methods.

asolntsev committed:
avoid using enhanced for loops

So called "enhanced for loops" allocate a new iterator, thus consuming more memory. In Netty, it's important to keep every single byte for better performance.

asolntsev committed:
add tests for all hash functions in ObjectUtil
